File: tests/iterative_expr_test.php
<?php
/*
 * OpenCReports test
 * Copyright (C) 2019-2024 Zoltán Böszörményi <zboszor@gmail.com>
 * See COPYING.LGPLv3 in the toplevel directory.
 */

require_once 'test_common.php';

$array = [
	[ "id", "name", "property", "age", "adult" ],
	[ "1", "Fred Flintstone", "strong", "31", "yes" ],
	[ "2", "Wilma Flintstone", "charming", "28", "yes" ],
	[ "3", "Pebbles Flintstone", "young", "5e-1", "no" ],
	[ "4", "Betty Rubble", "beautiful", "27", "yes" ],
	[ "5", "Barney Rubble", "small", "29", "yes" ]
];

$coltypes = [
	OpenCReport::RESULT_NUMBER, OpenCReport::RESULT_STRING, OpenCReport::RESULT_STRING, OpenCReport::RESULT_NUMBER, OpenCReport::RESULT_NUMBER
];

$o = new OpenCReport();

$ds = $o->datasource_add("array", "array");
$q = $ds->query_add("a", "array", "coltypes");

$e1 = $o->expr_parse("r.self + 2");

/* Initialize the value of "e1" to 0 */
$e1->init_results(OpenCReport::RESULT_NUMBER);
for ($i = 0; $i < OpenCReport::EXPR_RESULTS; $i++)
	$e1->set_nth_result_long($i, 0);
$e1->set_iterative_start_value(true);
$e1->resolve();
$e1->optimize();
echo "e1 expr (starts with initial value): "; flush();
$e1->print();

$e2 = $o->expr_parse("r.self + 2");
/* Initialize the value of "e2" to 0 */
$e2->init_results(OpenCReport::RESULT_NUMBER);
for ($i = 0; $i < OpenCReport::EXPR_RESULTS; $i++)
	$e2->set_nth_result_long($i, 0);
/* This iterative expression starts with evaluated value */
//ocrpt_expr_set_iterative_start_value(e2, true);
$e2->resolve();
$e2->optimize();
echo "e2 expr (starts with evaluated value): "; flush();
$e2->print();

echo "Both e1 and e2 were initialized to 0" . PHP_EOL . PHP_EOL;

$q->navigate_start();

while ($q->navigate_next()) {
	$r = $e1->eval();
	echo "e1 value: "; flush();
	$r->print();

	$r = $e2->eval();
	echo "e2 value: "; flush();
	$r->print();

	echo PHP_EOL;
}

$e1->free();
$e2->free();

File: tests/iterative_expr_test.c
/*
 * OpenCReports test
 * Copyright (C) 2019-2025 Zoltán Böszörményi <zboszor@gmail.com>
 * See COPYING.LGPLv3 in the toplevel directory.
 */

#include <stdio.h>

#include <opencreport.h>
#include "test_common.h"

#define ROWS 5
#define COLS 5
/* NOT static */ const char *array[ROWS + 1][COLS] = {
	{ "id", "name", "property", "age", "adult" },
	{ "1", "Fred Flintstone", "strong", "31", "yes" },
	{ "2", "Wilma Flintstone", "charming", "28", "yes" },
	{ "3", "Pebbles Flintstone", "young", "5e-1", "no" },
	{ "4", "Betty Rubble", "beautiful", "27", "yes" },
	{ "5", "Barney Rubble", "small", "29", "yes" }
};

/* NOT static */ const int32_t coltypes[5] = {
	OCRPT_RESULT_NUMBER, OCRPT_RESULT_STRING, OCRPT_RESULT_STRING, OCRPT_RESULT_NUMBER, OCRPT_RESULT_NUMBER
};

int main(int argc, char **argv) {
	opencreport *o = ocrpt_init();
	ocrpt_datasource *ds = ocrpt_datasource_add(o, "array", "array", NULL);
	ocrpt_query *q = ocrpt_query_add_data(ds, "a", (const char **)array, ROWS, COLS, coltypes, COLS);
	ocrpt_expr *e1, *e2;
	int i;

	e1 = ocrpt_expr_parse(o, "r.self + 2", NULL);
	/* Initialize the value of "e" to 0 */
	ocrpt_expr_init_results(e1, OCRPT_RESULT_NUMBER);
	for (i = 0; i < OCRPT_EXPR_RESULTS; i++)
		ocrpt_expr_set_nth_result_long(e1, i, 0);
	ocrpt_expr_set_iterative_start_value(e1, true);
	ocrpt_expr_resolve(e1);
	ocrpt_expr_optimize(e1);
	printf("e1 expr (starts with initial value): ");
	ocrpt_expr_print(e1);

	e2 = ocrpt_expr_parse(o, "r.self + 2", NULL);
	/* Initialize the value of "e" to 0 */
	ocrpt_expr_init_results(e2, OCRPT_RESULT_NUMBER);
	for (i = 0; i < OCRPT_EXPR_RESULTS; i++)
		ocrpt_expr_set_nth_result_long(e2, i, 0);
	/* This iterative expression starts with evaluated value */
	//ocrpt_expr_set_iterative_start_value(e2, true);
	ocrpt_expr_resolve(e2);
	ocrpt_expr_optimize(e2);
	printf("e2 expr (starts with evaluated value): ");
	ocrpt_expr_print(e2);

	printf("Both e1 and e2 were initialized to 0\n");
	printf("\n");

	ocrpt_query_navigate_start(q);

	while (ocrpt_query_navigate_next(q)) {
		ocrpt_result *r;

		r = ocrpt_expr_eval(e1);
		printf("e1 value: ");
		ocrpt_result_print(r);

		r = ocrpt_expr_eval(e2);
		printf("e2 value: ");
		ocrpt_result_print(r);

		printf("\n");
	}

	ocrpt_expr_free(e1);
	ocrpt_expr_free(e2);

	ocrpt_free(o);

	return 0;
}
